otter/convergence/selfheal.py

```diff
@@ -0,0 +1,169 @@
+"""
+Self heal service
+"""
+
+from effect import Effect
+from effect.do import do, do_return
+
+from kazoo.exceptions import LockTimeout
+from kazoo.protocol.states import KazooState
+
+from toolz.curried import filter
+
+from twisted.application.internet import TimerService
+from twisted.application.service import MultiService
+from twisted.internet.defer import inlineCallbacks, returnValue
+
+from txeffect import perform
+
+from otter.convergence.composition import tenant_is_enabled
+from otter.convergence.service import trigger_convergence
+from otter.log.intents import with_log
+from otter.models.intents import GetAllValidGroups, GetScalingGroupInfo
+from otter.models.interface import ScalingGroupStatus
+from otter.util.zk import GetChildren
+
+
+class SelfHeal(MultiService, object):
+    """
+    A service that triggers convergence on all the groups on interval basis.
+    Only one node is allowed to do this.
+    """
+
+    def __init__(self, dispatcher, kz_client, interval, log, clock,
+                 config_func):
+        super(SelfHeal, self).__init__()
+        self.disp = dispatcher
+        self.log = log.bind(otter_service="selfheal")
+        self.lock = kz_client.Lock("/selfheallock")
+        self.kz_client = kz_client
+        self.clock = clock
+        self.time_range = interval - 5
+        self.config_func = config_func
+        self.calls = []
+        timer = TimerService(
+            interval, lambda: self._converge_all().addErrback(self.log.err))
+        timer.clock = clock
+        timer.setServiceParent(self)
+
+    def stopService(self):
+        """
+        Stop service by cancelling any remaining scheduled calls and releasing
+        the lock
+        """
+        super(SelfHeal, self).stopService()
+        self._cancel_scheduled_calls()
+        return self.lock.release()
+
+    def _cancel_scheduled_calls(self):
+        """
+        Cancel any remaining scheduled calls
+        """
+        active = 0
+        for call in self.calls:
+            if call.active():
+                active += 1
+                call.cancel()
+        self.calls = []
+        return active
+
+    def health_check(self):
+        """
+        Return about whether this object has lock
+        """
+        d = is_lock_acquired(self.disp, self.lock)
+        return d.addCallback(lambda b: (True, {"has_lock": b}))
+
+    def _perform(self):
+        d = perform(self.disp, get_groups_to_converge(self.config_func))
+        d.addCallback(self._setup_converges)
+        return d.addErrback(self.log.err, "self-heal-err")
+
+    def _setup_converges(self, groups):
```

Comment:
stylistic, non-blocking:

This method is used only once, as a callback.  I generally prefer to avoid polluting the namespace of an object, even with private methods, if they're used only once as callbacks.  I'd define this as nested within `_perform` or even maybe make `_perform` an `inlineCallbacks` function.